### PR TITLE
[qt] Add map language preference menu

### PR DESCRIPTION
**Pull request by Ferenc-**

### CODE REVIEW

**Comment by Ferenc-** [comment]
Depends on https://github.com/organicmaps/organicmaps/pull/8237

**Comment by Ferenc-** [comment]
CC: @dmfrodrigues

File: android/app/src/main/java/app/organicmaps/settings/MapLanguageCode.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  {
+    return getCurrentMapLanguageCode();
+  }
+
+  public static void setMapLanguageCode(String locale)
+  {
+    setCurrentMapLanguageCode(locale);
+  }
+
+  public static void initializeCurrentMapLanguageCode()
```
Comment (biodranik) [inline android/app/src/main/java/app/organicmaps/settings/MapLanguageCode.java:21]: This one is not needed too, right?

File: android/app/src/main/java/app/organicmaps/settings/SettingsPrefsFragment.java [hunk 2]
```diff
@@ -71,6 +76,12 @@ private void updateVoiceInstructionsPrefsSummary()
     pref.setSummary(Config.TTS.isEnabled() ? R.string.on : R.string.off);
   }
 
+  private void updateMapLanguageCodeSummary()
+  {
+    final Preference pref = getPreference(getString(R.string.pref_map_locale));
+    pref.setSummary(String.valueOf(MapLanguageCode.getMapLanguageCode()));
```
Comment (biodranik) [inline android/app/src/main/java/app/organicmaps/settings/SettingsPrefsFragment.java:82]:
```suggestion
    pref.setSummary(MapLanguageCode.getMapLanguageCode());
```

File: android/app/src/main/java/app/organicmaps/settings/SettingsPrefsFragment.java [hunk 3]
```diff
@@ -104,6 +116,11 @@ else if (key.equals(getString(R.string.pref_help)))
       {
         startActivity(new Intent(requireActivity(), HelpActivity.class));
       }
+      else if (key.equals(getString(R.string.pref_map_locale)))
+      {
+        LanguagesFragment langFragment = (LanguagesFragment)getSettingsActivity().stackFragment(LanguagesFragment.class, getString(R.string.pref_map_locale_title), null);
+        langFragment.setListener(this);
```
Comment (biodranik) [inline android/app/src/main/java/app/organicmaps/settings/SettingsPrefsFragment.java:122]: @krozhdestvenski @rtsisyk is there a better way to set the listener?

File: android/app/src/main/res/xml/prefs_main.xml [hunk 4]
```diff
@@ -97,6 +97,12 @@
       android:summary="@string/enable_show_on_lock_screen_description"
       android:defaultValue="true"
       android:order="17"/>
+    <Preference
+        android:key="@string/pref_map_locale"
+        android:title="@string/pref_map_locale_title"
+        app:singleLineTitle="false"
+        android:persistent="false"
+        android:order="18"/>
```
Comment (biodranik) [inline android/app/src/main/res/xml/prefs_main.xml:105]: IS this option less important than other options? Should it be in a different position?

File: map/framework.hpp [hunk 5]
```diff
@@ -672,6 +672,8 @@ class Framework : public PositionProvider,
   void Save3dMode(bool allow3d, bool allow3dBuildings);
   void Load3dMode(bool & allow3d, bool & allow3dBuildings);
 
+  void SetMapLanguageCode(const std::string& languageCode);
```
Comment (biodranik) [inline map/framework.hpp:675]:
```suggestion
  void SetMapLanguageCode(std::string const & languageCode);
```

File: platform/preferred_languages.cpp [hunk 6]
```diff
@@ -1,4 +1,7 @@
 #include "platform/preferred_languages.hpp"
+#include "platform/settings.hpp"
```
Comment (biodranik) [inline platform/preferred_languages.cpp:2]: Settings should be checked in framework.cpp

File: platform/preferred_languages.cpp [hunk 7]
```diff
@@ -169,6 +172,41 @@ std::string GetCurrentNorm()
   return Normalize(GetCurrentOrig());
 }
 
+std::string GetUtf8MultilangSupported()
+{
+  std::string languageCode;
+
+  auto const & systemLanguages = GetSystemPreferred();
+  for (const auto & systemLanguage: systemLanguages)
```
Comment (biodranik) [inline platform/preferred_languages.cpp:180]:
```suggestion
  for (auto const & systemLanguage : GetSystemPreferred())
```

File: platform/preferred_languages.cpp [hunk 8]
```diff
@@ -169,6 +172,41 @@ std::string GetCurrentNorm()
...
+  {
+    const std::string normalizedLang = Normalize(systemLanguage);
```
Comment (biodranik) [inline platform/preferred_languages.cpp:182]:
```suggestion
    std::string const normalizedLang = Normalize(systemLanguage);
```

File: platform/preferred_languages.cpp [hunk 9]
```diff
@@ -169,6 +172,41 @@ std::string GetCurrentNorm()
...
+    if (StringUtf8Multilang::GetLangIndex(normalizedLang) != StringUtf8Multilang::kUnsupportedLanguageCode)
+    {
+      languageCode = normalizedLang;
+      break;
+    }
```
Comment (biodranik) [inline platform/preferred_languages.cpp:187]:
```suggestion
    return normalizedLang;
```

File: platform/preferred_languages.cpp [hunk 10]
```diff
@@ -169,6 +172,41 @@ std::string GetCurrentNorm()
...
+  }
+
+  if (languageCode.empty())
+    throw std::runtime_error("Unsupported language.");
+
+  return languageCode;
```
Comment (biodranik) [inline platform/preferred_languages.cpp:193]:
```suggestion
  return "en";
```

File: platform/preferred_languages.cpp [hunk 11]
```diff
@@ -169,6 +172,41 @@ std::string GetCurrentNorm()
...
+}
+
+std::string GetCurrentMapLanguageCode()
+{
+  std::string languageCode;
+  if (!settings::Get(settings::kMapLanguageCode, languageCode) || languageCode.empty())
+    languageCode = GetUtf8MultilangSupported();
```
Comment (biodranik) [inline platform/preferred_languages.cpp:200]: 1. The name GetUtf8MultilangSupported is not great.
2. The code of GetUtf8MultilangSupported function can be directly placed here.

File: platform/preferred_languages.cpp [hunk 12]
```diff
@@ -169,6 +172,41 @@ std::string GetCurrentNorm()
...
+  
+  return languageCode;
+}
+
+void SetMapLanguageCode(std::string const & languageCode)
+{
+  settings::Set(settings::kMapLanguageCode, languageCode);
+}
```
Comment (biodranik) [inline platform/preferred_languages.cpp:208]: This should be done in framework.

File: qt/preferences_dialog.cpp [hunk 13]
```diff
@@ -97,6 +99,28 @@ namespace qt
       });
     }
 
+    QLabel * mapLanguageLabel = new QLabel("Map Language");
+    QComboBox * mapLanguageComboBox = new QComboBox();
+    {
+      StringUtf8Multilang::Languages const & supportedLanguages = StringUtf8Multilang::GetSupportedLanguages();
+      QStringList languagesList = QStringList();
+      for (auto const & language : supportedLanguages)
```
Comment (biodranik) [inline qt/preferences_dialog.cpp:107]: I switched languages in the desktop version and found a bug that incorrect language is set for some languages (e.g. Polish/Arabic, and other in the list below). Looks like the offset/lang code was shifted by 1 for some reason.

I didn't test it on Android, maybe the same issue is there too.

File: qt/preferences_dialog.cpp [hunk 14]
```diff
@@ -97,6 +99,28 @@ namespace qt
...
+        languagesList << QString::fromStdString(std::string(language.m_name));
+
+      mapLanguageComboBox->addItems(languagesList);
+      mapLanguageComboBox->setCurrentIndex(StringUtf8Multilang::GetLangIndex(languages::GetCurrentMapLanguageCode()));
+      connect(mapLanguageComboBox, &QComboBox::currentIndexChanged, [&framework](int index)
+      {
+        auto const & mapLanguageCode = std::string(StringUtf8Multilang::GetLangByCode(index));
+        if (mapLanguageCode.empty()) // Empty means unsupported
```
Comment (biodranik) [inline qt/preferences_dialog.cpp:115]: Aren't all languages retrieved by `StringUtf8Multilang::GetSupportedLanguages()` supported? Isn't this check redundant and may be removed or replaced by the ASSERT?

**Review by biodranik — CHANGES_REQUESTED** [review]
@Ferenc- @krozhdestvenski thanks! I left comments there that should be fixed.

C++ changes, Android changes, strings regeneration and Qt changes should be later moved into separate commits before the merge.

File: platform/preferred_languages.cpp (hunk 10 above)
Comment (Ferenc-) [reply platform/preferred_languages.cpp:193]: If we want a default value, then [shoudn't that be](https://github.com/organicmaps/organicmaps/blob/master/coding/string_utf8_multilang.cpp#L14) `"default"`?

File: qt/preferences_dialog.cpp (hunk 13 above)
Comment (Ferenc-) [reply qt/preferences_dialog.cpp:107]: This here was qt specific. Fixed.

File: platform/preferred_languages.cpp (hunk 7 above)
Comment (Ferenc-) [reply platform/preferred_languages.cpp:180]: Done

File: platform/preferred_languages.cpp (hunk 8 above)
Comment (Ferenc-) [reply platform/preferred_languages.cpp:182]: done

File: platform/preferred_languages.cpp (hunk 9 above)
Comment (Ferenc-) [reply platform/preferred_languages.cpp:187]: done

File: platform/preferred_languages.cpp (hunk 10 above)
Comment (Ferenc-) [reply platform/preferred_languages.cpp:193]: As in `return StringUtf8Multilang::kDefaultCode;`

File: platform/preferred_languages.cpp (hunk 11 above)
Comment (Ferenc-) [reply platform/preferred_languages.cpp:200]: removed & done

File: map/framework.hpp (hunk 5 above)
Comment (Ferenc-) [reply map/framework.hpp:675]: done

File: android/app/src/main/java/app/organicmaps/settings/SettingsPrefsFragment.java (hunk 2 above)
Comment (Ferenc-) [reply android/app/src/main/java/app/organicmaps/settings/SettingsPrefsFragment.java:82]: done

File: platform/preferred_languages.cpp (hunk 12 above)
Comment (Ferenc-) [reply platform/preferred_languages.cpp:208]: done

File: android/app/src/main/java/app/organicmaps/editor/Editor.java [hunk 15]
```diff
@@ -95,8 +96,9 @@ public static void nativeSetBuildingLevels(String levels)
   public static native NamesDataSource nativeGetNamesDataSource();
   public static native void nativeSetNames(@NonNull LocalizedName[] names);
   public static native LocalizedName nativeMakeLocalizedName(String langCode, String name);
+  //public static native Language[] nativeGetSupportedLanguages();
```
[the remark this answers is not in the document]
Comment (Ferenc-) [reply android/app/src/main/java/app/organicmaps/editor/Editor.java:99]: removed

File: android/app/src/main/cpp/app/organicmaps/editor/Editor.cpp [hunk 16]
```diff
@@ -293,6 +293,12 @@ Java_app_organicmaps_editor_Editor_nativeGetSupportedLanguages(JNIEnv * env, jcl
                           });
 }
 
+JNIEXPORT jboolean JNICALL
+Java_app_organicmaps_editor_Editor_nativeIsServiceLanguage(JNIEnv * env, jclass clazz, jstring lang)
+{
+    return StringUtf8Multilang::IsServiceLang(jni::ToNativeString(env, lang));
```
[the remark this answers is not in the document]
Comment (Ferenc-) [reply android/app/src/main/cpp/app/organicmaps/editor/Editor.cpp:299]: fixed

File: android/app/src/main/cpp/app/organicmaps/settings/MapLanguageCode.cpp [hunk 17]
```diff
@@ -0,0 +1,18 @@
+#include "app/organicmaps/Framework.hpp"
+
+#include "platform/settings.hpp"
+
+extern "C"
+{
+  JNIEXPORT void JNICALL
```
[the remark this answers is not in the document]
Comment (Ferenc-) [reply android/app/src/main/cpp/app/organicmaps/settings/MapLanguageCode.cpp:7]: Removed indentation

File: platform/preferred_languages.cpp (hunk 6 above)
Comment (Ferenc-) [reply platform/preferred_languages.cpp:2]: done

File: platform/preferred_languages.cpp [hunk 18]
```diff
@@ -169,6 +172,41 @@ std::string GetCurrentNorm()
...
+  if (languageCode.empty())
+    throw std::runtime_error("Unsupported language.");
+
+  return languageCode;
+}
+
+std::string GetCurrentMapLanguageCode()
+{
+  std::string languageCode;
+  if (!settings::Get(settings::kMapLanguageCode, languageCode) || languageCode.empty())
```
[the remark this answers is not in the document]
Comment (Ferenc-) [reply platform/preferred_languages.cpp:199]: moved

File: android/app/src/main/java/app/organicmaps/settings/MapLanguageCode.java [hunk 19]
```diff
@@ -0,0 +1,26 @@
+package app.organicmaps.settings;
+
+public class MapLanguageCode
+{
+  private static native String getDefaultMapLanguageCode();
```
[the remark this answers is not in the document]
Comment (Ferenc-) [reply android/app/src/main/java/app/organicmaps/settings/MapLanguageCode.java:5]: removed

File: android/app/src/main/java/app/organicmaps/editor/LanguagesFragment.java [hunk 20]
```diff
@@ -24,28 +25,43 @@ public interface Listener
...
+            : new HashSet<>();
 
     List<Language> languages = new ArrayList<>();
     for (Language lang : Editor.nativeGetSupportedLanguages())
     {
-      if (!existingLanguages.contains(lang.code))
-        languages.add(lang);
+      if (existingLanguages.contains(lang.code))
+        continue;
+      if (Editor.nativeIsServiceLanguage(lang.code))
```
[the remark this answers is not in the document]
Comment (biodranik) [reply android/app/src/main/java/app/organicmaps/editor/LanguagesFragment.java:44]: Do you have better ideas?

**Comment by biodranik** [comment]
Issues with simplified Chinese: https://omaps.app/4kLRTzEIJz/

An empty line is displayed under the icon on Android
telegram-cloud-photo-size-1-5183877724018682986-w

**Comment by biodranik** [comment]
On iOS there's a different issue: enabling Simplified Chinese for the app or in the system always displays the name in English.

**Comment by Ferenc-** [comment]
These are valid points, and perhaps they should be tracked in separate issues, because this PR is realy only about the qt version.

**Comment by biodranik** [comment]
Uuups, sorry, I meant this PR: https://github.com/organicmaps/organicmaps/pull/8237#issuecomment-2198638786

**Comment by pastk** [comment]
I've tested it on my desktop and it works well!

Though Place Page language is not affected (shows everything in En or transliterated), but it could be improved later.

**Comment by biodranik** [comment]
> Though the Place Page language is not affected (shows everything in En or transliterated), it could be improved later.

UI language and map language are two different things. There could be a UI language that does not have a corresponding map language.

**Comment by pastk** [comment]
"UI language" is our "hardcoded" UI strings (strings.txt ...)

Whereas I referred to e.g. names that are part of map data, but are displayed in PP, e.g.
image

**Comment by pastk** [comment]
Android version PP and search results are also affected (the lang doesn't change there).

**Comment by Ferenc-** [comment]
Closing in favor of https://github.com/organicmaps/organicmaps/pull/9618

**Pull request closed without merging**